src/metabase/query_processor/middleware/annotate.clj

```diff
@@ -517,10 +517,14 @@
 (defn- flow-field-metadata
   "Merge information about fields from `source-metadata` into the returned `cols`."
   [source-metadata cols dataset?]
-  (let [index           (fn [col] (or (:id col) (:name col "")))
-        index->metadata (m/index-by index source-metadata)]
-    (for [col cols]
-      (if-let [source-metadata-for-field (-> col index index->metadata)]
+  (let [by-key (m/index-by (comp qp.util/field-ref->key :field_ref) source-metadata)]
```

Comment:
Part of this is reverse the decision of using "name" as index we made in #19437

Reply:
Note that we could still keep that strategy of using `id then name` as an index and it'll still work.
But I think since using `field_ref` works perfectly and we also use `field_ref` as an index in `qp.util/combine-metadata`, so I think I'll keep this one consistent by reverting to the previous strategy.